### ISSUE TITLE
Redis master-slave replication

### ISSUE THREAD

**Issue by nlsun** [tillkruss/Redis-Object-Cache#26]
Support for Redis master-slave replication. To implement it this plugin needs to additionally direct reads and writes to different redis servers.

**Comment by tillkruss** [comment]
I was thinking about this the last couple if days. Only Predis supports sharing and replication, HHVM and PHPRedis do not.

Could you tell me more about what exactly you had in mind?

**Comment by nlsun** [comment]
I was think that the burden of configuring replication would be managed by the user.

I'm not entirely sure how this works but I think it would be sufficient if you could just configure in this plugin "If I do a write go to this server" and "If I do a read go to this server" then that will be enough.

**Comment by tillkruss** [comment]
Okay, so only replication, no clustering or client sharding?

**Comment by nlsun** [comment]
I think replication is alot easier as a first step. Not sure what everyone is using in their own setups but i'd image master-slave is a pretty common scenario.

**Comment by tillkruss** [comment]
I was thinking to make PHP7 a requirement for replication. If people care about performance they are probably using PHP7 and the plugins could reply on array constants.

```php
define( 'WP_REDIS_SERVERS', ['master' => 'foobar', 'slaves' => ['foo', 'bar']] );
```

**Comment by nlsun** [comment]
Also, I was checking out Predis (https://github.com/nrk/predis) and it seems like they have support for master-slave replication

**Comment by nlsun** [comment]
actually, is it an issue if you only offer these features (redis cluster, redis replication) if predis is selected?

**Comment by tillkruss** [comment]
No, I was thinking to make replication only work with Predis and PHP7.

**Comment by esobczak1970** [comment]
I think it would be wise to allow for configuring an unlimited number of redis servers. And then being able to choose one write target and one read target. Beyond that it is my understanding that to do the clustering for example you'd implement haproxy making it the target of read or write or both. The haproxy would then switch between the actual redis servers based on availability and load.

By defining some number of redis servers you could then also designate a failover as another alternative where let's say your primary target goes down. On an info check if it doesn't get X responses it would failover to an alternate target/server. Additionally there are lots of other reasons to be able to define all servers of haproxied targets. So it is a good idea to be able to configure these now even if the functionality is not there to fully take advantage in all ways.

Eventually the plugin could promote slaves, retrieve info on each server, and a whole list of other features.

Love the idea of having the PHP7 optimization available, but would be nice if older PHP could still be supported. I know we are in the process of transitioning but haven't done it yet.

**Comment by nlsun** [comment]
I actually think stuff like slave promotion all other other features should be handled out of band. For example, there is Redis Sentinel. This plugin should be kept as simple as possible and just be really good at reading and writing. Where those reads and writes actually go can be configured through other methods (eg. haproxy).

**Comment by esobczak1970** [comment]
If I want writes for example to go to a master and reads to go to local host on my app servers of which there are 3 so I would have 3 slaves. There is not a way to configure this. And even if I configure haproxy it seems to me it could be used to balance for example 3 slaves for reading but they would also have to accept writing which would need replication to do any good. And yes in some ways the autopromotion from slave to master on failure would be better handled perhaps by sentinel. This would give you just another level of control and flexibility and as I mentioned that would only be one reason for having the servers delineated.

**Comment by tillkruss** [comment]
Maybe one of you could write a couple of common scenarios and how the Predis configuration would look?

**Comment by nlsun** [comment]
But we can solve that with 2 simple configuration options: read_host and write_host. So in your scenario you would write to your master and read from localhost

**Comment by nlsun** [comment]
I want to know how predis supports clustering and replication before making such a design decision though.

**Comment by tillkruss** [comment]
See Replication and Cluster: https://github.com/nrk/predis#aggregate-connections

**Comment by esobczak1970** [comment]
I'll put together a few scenarios when I get a chance.

Eric

-----Original Message-----

**Comment by esobczak1970** [comment]
Correct that is why we need the separate read and write. Though if we want to be able to get stats of an info summary from different servers it would come in handy to be able to configure them all.

But your right at least having that at the plugin level would help greatly.

Eric

-----Original Message-----

**Comment by tillkruss** [comment]
Configure a cluster:
```php
define( 'WP_REDIS_CLUSTER', [
    'tcp://10.0.0.1:6379?database=10&alias=one',
    'tcp://10.0.0.2:6379?database=15&alias=two'
] );
```

Configure Master-Slave:
```php
define( 'WP_REDIS_SERVERS', [
    'tcp://10.0.0.1:6379?database=10&alias=master',
    'tcp://10.0.0.2:6379?database=15&alias=slave-01',
    'tcp://10.0.0.2:6379?database=15&alias=slave-02'
] );
```

Thoughts?

**Comment by nlsun** [comment]
What does the `database=10` part mean?

**Comment by tillkruss** [comment]
The numeric id of the database. 10 is just an example.

**Comment by tillkruss** [comment]
Let me know if it works:
https://github.com/tillkruss/Redis-Object-Cache/tree/aggregates

**Comment by esobczak1970** [comment]
So our host has this configured for us at present

Redis is now running as master-slave on db1-new and db2-new.

With shut down of redis on db1-new, the redis instance on db2-new should automatically be promoted to master.

It works in conjunction with HAproxy on each web server, so please utilize the hostname of `bal` for redis operations.

Using the bal host will direct the connections to the nearest HAproxy instance where:

host `bal` port `6379` is to be used for all writes (always directs toward the current redis master).
host `bal` port `6380` is to be used for reads (balances connections between slave and master).

Would you please let us know once all code configurations have been changed to reflect the `bal` hostname, rather than db1-new (or 192.168.194.11)?

host `bal` on the dev server points to the local redis instance there, so there shouldn't be a need to change your code's configuration files between production and staging.


So you'll notice part of the key in our current config is to support an alternate read and write port. But most of what you describe sounds right.

Eric


-----Original Message-----

**Comment by tillkruss** [comment]
Can you please make sure that the Predis library supports that setup?

Also, check out the `aggregates` branch. I’ve already implemented clusters and replication. Please test it and feel free to submit PRs.

Cheers,
Till

**Comment by esobczak1970** [comment]
Predis just facilitates the connection via PHP I think it's really in the object-cache.php that you would facilitate a trigger to write at one target versus another.

Eric

-----Original Message-----

**Comment by tillkruss** [comment]
Please read the Predis docs.

**Comment by tillkruss** [comment]
Replication and clustering will be included in `v1.3`.
